In [2]:
import pandas as pd

ler todos os arquivos, em formato de texto

In [3]:
customers = pd.read_csv("../data/raw/customers.csv", dtype=str)
products = pd.read_csv("../data/raw/products.csv", dtype=str)
orders = pd.read_csv("../data/raw/orders.csv", dtype=str)
order_items = pd.read_csv("../data/raw/order_items.csv", dtype=str)
reviews = pd.read_csv("../data/raw/reviews.csv", dtype=str)

In [4]:
tables = {
    "customers": customers,
    "products": products,
    "orders": orders,
    "order_items": order_items,
    "reviews": reviews,
}

quantas linhas e colunas tem cada tabela?

In [5]:
for name, df in tables.items():
    print(f"\n{name}")
    print(df.shape)


customers
(3060, 11)

products
(606, 7)

orders
(15088, 8)

order_items
(26191, 6)

reviews
(8080, 6)


olhar os dados de verdade, nao confiar só nos tipos

In [6]:
for name, df in tables.items():
    print("="*60)
    print(name)
    display(df.head(10))

customers


,customer_id,first_name,last_name,email,phone,cpf,birth_date,city,state,signup_date,customer_segment
0,2051,Lunna,Farias,leticia29@example.com,(71)491351387,268.415.739-19,1991-10-19T00:00:00,Mendonça,GO,04/11/2024 00:00,NOVO
1,2229,Srta.,Letícia Martins,kamilly12@example.net,(87)792142376,61475982364,1987-10-02T00:00:00,Caldeira,RS,22/11/2025,inativo
2,1366,José,Miguel Nascimento,guilhermecorreia@example.org,96-760428699,259.430.167-16,15/09/1991 00:00,Farias das Pedras,mg,2023-10-03T00:00:00,NOVO
3,1449,Dr.,Henrique Garcia,apolloborges@example.org,56748263098,098.465.173-00,05/10/1978,das Neves,am,02/12/2024 00:00,vip
4,1836,Davi,Lucas Silveira,abreuenzo-gabriel@example.org,(90) 88785-0083,NaN,2003-07-30T00:00:00,da Luz de Goiás,df,05/26/2025,novo
5,731,Jade,Camargo,kpacheco@example.net,+55 15 68148-3989,743.805.261-35,30/03/1998 00:00,Leão do Norte,rs,2025-01-13,vip
6,1101,Sr.,Fernando Cassiano,pachecocaroline@example.com,(64)284275160,684.527.103-80,03/23/1961,Machado,sp,18/10/2023,NOVO
7,2139,Dr.,Davi Lucas Ribeiro,NaN,50253088660,103.624.578-08,07-12-1981,Ribeiro,PR,10/09/2024,NaN
8,405,Raul,Santos,rpires@example.net,+55 86 28623-7999,NaN,26/12/1986 00:00,Aparecida dos Dourados,go,01/03/2025,inativo
9,1949,Diogo,Silveira,isabela26@example.com,NaN,132.097.586-03,04-10-1960,Macedo do Norte,df,13/01/2026,vip


products


,product_id,product_name,category,brand,price,stock_quantity,created_at
0,155,Quadro Decorativo 256 Genérico,decoracao,Genérico,"8983,68",-1,2025/02/13
1,13,Monitor 1pol Multilaser,INFORMATICA,Multilaser,"1208,34",-1,08/02/2025
2,545,Máquina de Lavar 256Kg Brastemp,eletrodomesticos,Brastemp,1428.71,-1,NaN
3,2,Soundbar 8 AOC,Eletrônicos,AOC,2350.79,4,03-04-2024
4,577,Fone de Ouvido Bluetooth 1 AOC,eletronicos,AOC,8307.74,NaN,NaN
5,113,Smart TV 4pol Sony,Eletrônicos,Sony,"R$ 1.590,09",-1,03/24/2025
6,50,Rack para TV 55 Madesa,Moveis,Madesa,5938.53,95,0000-00-00
7,484,Cama Box Casal 1 Artely,Móveis,Artely,NaN,160,2023-08-08T00:00:00
8,89,Cama Box Casal 50 Politorno,Móveis,Politorno,8773.17,-1,14/06/2026
9,515,Geladeira Frost Free 50L LG,Eletrodomésticos,LG,"R$ 7.193,60",-5,08/13/2023


orders


,order_id,customer_id,order_date,status,payment_method,shipping_city,shipping_state,total_amount
0,14062,1424,2024-10-23,entregue,boleto,Vargas,GO,3668.29
1,4000,801,13/06/2025,Entregue,Pix,Aparecida Verde,BA,"R$ 5.966,18"
2,5000,1591,10/25/2025,shipped,boleto,Correia Grande,SP,"5143,41"
3,1531,1254,2025-08-06T18:40:21,cancelled,Pix,Andrade do Norte,PE,"17538,0"
4,6555,1274,09/02/2025,Entregue,boleto,Rezende de Goiás,RS,1463.6
5,4479,1823,2025-11-24T21:24:07,shipped,Pix,Leão de das Neves,GO,7465.22
6,7646,2429,2025-06-25,refunded,paypal,Farias,GO,21914.23
7,6361,2463,2024-10-03,Pendente,cartao_credito,Costela,AM,"R$ 781,18"
8,1283,2728,2025/04/16,shipped,boleto,da Rocha,RJ,"7130,46"
9,4755,2626,11/07/2025,Entregue,paypal,Campos do Norte,RJ,5950.33


order_items


,order_item_id,order_id,product_id,quantity,unit_price,discount_pct
0,17510,10019,80,1,1378.46,0
1,20948,11986,190,1,2603.05,0
2,12803,7333,419,1,2283.85,0.05
3,8107,4630,63,1,"R$ 2.070,54",0.1
4,399,237,218,1,1858.16,NaN
5,16687,9550,29,3,5246.72,0.1
6,3464,1981,86,1,"R$ 8.570,63",0
7,23636,13535,394,2,5702.23,0
8,9060,5172,192,1,4992.93,0.05
9,24376,13957,4,1,"R$ 5.693,68",0


reviews


,review_id,product_id,customer_id,rating,review_text,review_date
0,2173,97,2142,5,"Poderia ser melhor, mas cumpre o que promete.",25/11/2025
1,845,487,1942,4,"Ótimo produto, super indico!",0000-00-00
2,4742,421,444,3,"Excelente custo benefício, recomendo.",2026/07/07
3,2858,350,1634,4,Entrega demorou mais do que o esperado.,2025-08-22
4,1762,239,2883,5,Atendeu minhas expectativas.,08/01/2025 15:21
5,7781,217,498,4,"Ótimo produto, super indico!",0000-00-00
6,3888,312,1787,5,Não gostei da qualidade do material.,07/18/2025
7,7896,304,2357,3,"Veio com defeito, tive que trocar.",2025/11/27
8,4396,193,1658,3,"Excelente custo benefício, recomendo.",0000-00-00
9,4515,40,1360,5,"Produto muito bom, chegou rápido!",2024-08-30


 quantos nulos por coluna?

In [7]:
for name, df in tables.items():
    print("="*60)
    print(name)
    print(df.isnull().sum())

customers
customer_id           0
first_name            0
last_name             0
email               145
phone               256
cpf                 300
birth_date          148
city                  0
state                 0
signup_date         159
customer_segment    487
dtype: int64
products
product_id          0
product_name        0
category            0
brand               0
price              22
stock_quantity    135
created_at         37
dtype: int64
orders
order_id             0
customer_id          0
order_date         787
status               0
payment_method    1969
shipping_city        0
shipping_state       0
total_amount      1298
dtype: int64
order_items
order_item_id       0
order_id            0
product_id          0
quantity            0
unit_price       1285
discount_pct     1320
dtype: int64
reviews
review_id         0
product_id        0
customer_id       0
rating          400
review_text    1630
review_date     391
dtype: int64


LINHAS DUPLICADAS

In [8]:
for name, df in tables.items():
    print(f"{name}: {df.duplicated().sum()}")

customers: 1
products: 0
orders: 88
order_items: 0
reviews: 80


IDS DUPLICADOS
 se df.duplicated().sum() for igual a df['customer_id'].duplicated().sum(), os duplicados são cópias exatas
 mas se df['customer_id'].duplicated().sum() for maior, significa que existe o mesmo customer_id aparecendo com dados diferentes (ex.: email diferente em cada linha)

In [9]:
id_columns = {
    "customers": ["customer_id"],
    "products": ["product_id"],
    "orders": ["order_id", "customer_id"],
    "order_items": ["order_item_id", "order_id", "product_id"],
    "reviews": ["review_id", "product_id", "customer_id"]
}

for table_name, cols in id_columns.items():
    print("="*60)
    print(table_name)

    df = tables[table_name]

    for col in cols:
        print(col, "->", df[col].duplicated().sum())
    

customers
customer_id -> 60
products
product_id -> 6
orders
order_id -> 88
customer_id -> 11894
order_items
order_item_id -> 0
order_id -> 11191
product_id -> 25453
reviews
review_id -> 80
product_id -> 7480
customer_id -> 5308


VER colunas categóricas

In [10]:
categorical_columns  = {
    "customers": ["state", "customer_segment"],
    "products": ["brand","category"],
    "orders": ["status", "payment_method", "shipping_state"]
}

for table_name, cols in categorical_columns.items():

    df = tables[table_name]

    for col in cols:
        print("="*60)
        print(table_name, "-", col)

        print(df[col].value_counts(dropna=False))

customers - state
state
df    156
sp    151
DF    145
PR    143
SP    134
BA    133
SC    133
PE    133
AM    130
ba    129
ce    129
CE    126
mg    124
pr    124
RS    123
pe    123
MG    121
sc    121
go    118
GO    114
rj    114
am    112
rs    112
RJ    112
Name: count, dtype: int64
customers - customer_segment
customer_segment
vip        526
regular    519
novo       512
NOVO       510
inativo    506
NaN        487
Name: count, dtype: int64
products - brand
brand
Samsung         61
Genérico        46
LG              46
Multilaser      39
Kappesberg      35
Casa Store      34
Trinity         33
Consul          32
Electrolux      27
Dell            26
Artely          25
Móveis Bento    23
Sony            22
Philips         22
Acer            21
Lenovo          21
AOC             18
Madesa          17
Politorno       17
Rudnick         15
Brastemp        13
Positivo        13
Name: count, dtype: int64
products - category
category
Eletrodomésticos    50
informatica         43
eletro

ver os formatos das datas utilizados

In [11]:
date_columns = {
    "customers": ["birth_date", "signup_date"],
    "products":  ["created_at"],
    "orders":    ["order_date"],
    "reviews":   ["review_date"]
}

for table_name, cols in date_columns.items():

    df = tables[table_name]

    for col in cols:
        print("="*60)
        print(table_name, "-", col)

        print(df[col].dropna().unique()[:20])

customers - birth_date
<ArrowStringArray>
['1991-10-19T00:00:00', '1987-10-02T00:00:00',    '15/09/1991 00:00',
          '05/10/1978', '2003-07-30T00:00:00',    '30/03/1998 00:00',
          '03/23/1961',          '07-12-1981',    '26/12/1986 00:00',
          '04-10-1960',          '1979-07-03',          '08/06/1996',
    '14/11/1962 00:00',          '2001/11/12',          '0000-00-00',
          '04-12-1962',          '01/13/1992',          '06/15/1988',
          '01-12-1969',          '1981-10-06']
Length: 20, dtype: str
customers - signup_date
<ArrowStringArray>
[   '04/11/2024 00:00',          '22/11/2025', '2023-10-03T00:00:00',
    '02/12/2024 00:00',          '05/26/2025',          '2025-01-13',
          '18/10/2023',          '10/09/2024',          '01/03/2025',
          '13/01/2026',          '08/10/2023',    '03/11/2025 00:00',
          '28/05/2025',          '28/12/2022',    '27/12/2023 00:00',
          '01-09-2023',          '16-11-2022',          '20-07-2022',
     

Amostra de formatos de dinheiro

In [12]:
price_columns = {
    "products":  ["price"],
    "order_items":  ["unit_price"],
    "orders":   ["total_amount"]
}

for table_name, cols in price_columns.items():

    df = tables[table_name]

    for col in cols:
        print("="*60)
        print(table_name, "-", col)

        print(df[col].dropna().unique()[:20])

products - price
<ArrowStringArray>
[    '8983,68',     '1208,34',     '1428.71',     '2350.79',     '8307.74',
 'R$ 1.590,09',     '5938.53',     '8773.17', 'R$ 7.193,60',     '1801.87',
     '4525.27',       '829.9',      '388,07',     '1017.14', 'R$ 8.657,04',
     '2381,35', 'R$ 1.762,22',     '5969.58',     '2781.77',     '1236.53']
Length: 20, dtype: str
order_items - unit_price
<ArrowStringArray>
[    '1378.46',     '2603.05',     '2283.85', 'R$ 2.070,54',     '1858.16',
     '5246.72', 'R$ 8.570,63',     '5702.23',     '4992.93', 'R$ 5.693,68',
     '3161.29',   'R$ 430,41',     '8469.86',     '6649.54',     '8635.23',
     '2911.79',     '7651.28', 'R$ 5.287,09',      '470.69', 'R$ 5.262,36']
Length: 20, dtype: str
orders - total_amount
<ArrowStringArray>
[     '3668.29',  'R$ 5.966,18',      '5143,41',      '17538,0',
       '1463.6',      '7465.22',     '21914.23',    'R$ 781,18',
      '7130,46',      '5950.33', 'R$ 20.432,25', 'R$ 28.781,55',
       '6788.3', 'R$ 15.992,12

integridade referencial

ver orders com costumer id que nao tem na tabela de costumers
ver orders items id com products ids que nao existem na tabela de products

In [13]:
customers_ids = set(customers["customer_id"])
orders_ids = set(orders["customer_id"])

missing = orders_ids - customers_ids

print("Clientes inexistentes:", len(missing))

products_ids = set(products["product_id"])
items_ids = set(order_items["product_id"])

missing = items_ids - products_ids

print("Produtos inexistentes:", len(missing))

Clientes inexistentes: 210
Produtos inexistentes: 138


## RESUMO FINAL DOS ACHADOS — mapa de sujeira do dataset

### customers (3.060 linhas)
- **PK duplicada:** 60 `customer_id` repetidos (1 linha 100% idêntica; os demais têm o mesmo ID com dados diferentes — precisa decidir qual registro manter na Silver).
- **Nulos relevantes:** `customer_segment` (487), `phone` (256), `cpf` (300), `signup_date` (159), `birth_date` (148), `email` (145).
- **Datas (`birth_date`, `signup_date`):** pelo menos 5 formatos misturados — ISO com hora (`1991-10-19T00:00:00`), `dd/mm/yyyy` com e sem hora `00:00`, `dd-mm-yyyy`, `yyyy/mm/dd`, `mm/dd/yyyy`, e o sentinela inválido `0000-00-00`.
- **Categóricas sujas:** `state` com maiúsc/minúsc misturado (`df`/`DF`, `sp`/`SP`...); `customer_segment` com `NOVO`/`novo` duplicando o mesmo valor.
- **Outros:** `first_name`/`last_name` mal segmentados — títulos como "Sr.", "Srta.", "Dr." aparecem em `first_name` no lugar do nome; `phone` e `cpf` com e sem máscara/formatação.

### products (606 linhas)
- **PK duplicada:** 6 `product_id` repetidos.
- **Nulos relevantes:** `stock_quantity` (135), `created_at` (37), `price` (22).
- **Sentinela inválido:** `stock_quantity` com valores negativos (ex.: `-1`, `-5`) — estoque não pode ser negativo, é um marcador de "sem dado" disfarçado de número.
- **Dinheiro (`price`):** 3 formatos — `1428.71` (ponto decimal), `"8983,68"` (vírgula BR), `"R$ 1.590,09"` (moeda BR completa com separador de milhar).
- **Datas (`created_at`):** mesma mistura de formatos de `customers`, incluindo `0000-00-00`.
- **Encoding/categoria (`category`):** a mesma categoria escrita em até 5 grafias diferentes por causa de caixa E mojibake de encoding — ex. "eletrônicos": `Eletrônicos`, `ELETRONICOS`, `eletronicos`, `Eletronicos`, `Eletrônicos` (UTF-8 lido como Latin-1). Mesmo padrão em "informática", "decoração" e "móveis".

### orders (15.088 linhas)
- **PK duplicada:** 88 `order_id` repetidos (linhas 100% idênticas).
- **FK órfã:** 210 `customer_id` que não existem em `customers` — pedidos "fantasma" sem cliente.
- **Nulos relevantes:** `payment_method` (1.969), `total_amount` (1.298), `order_date` (787).
- **Categórica mais suja do dataset (`status`):** ~24 variações do mesmo conjunto de 5 estados reais, misturando português/inglês, maiúsc/minúsc e até espaço em branco à direita (`entregue` aparece 2x com contagens diferentes por causa disso). Precisa de um de-para explícito PT↔EN + normalização de caixa/espaço.
- **`payment_method`:** 3 grafias para "cartão de crédito" (`cartao credito`, `CARTAO_CREDITO`, `cartao_credito`) e caixa mista em `Pix`/`pix`.
- **Datas (`order_date`):** mesmos formatos misturados, incluindo timestamp completo (`2025-08-06T18:40:21`) e o formato americano inequívoco `10/25/2025` (dia 25 prova que não pode ser `dd/mm`).
- **Dinheiro (`total_amount`):** mesmos 3 formatos de `price`.

### order_items (26.191 linhas)
- **PK limpa:** `order_item_id` sem nenhum duplicado — a única chave primária 100% íntegra do dataset.
- **FK órfã:** 138 `product_id` (na faixa 601–800) que não existem em `products` (que só vai até 600).
- **Nulos relevantes:** `unit_price` (1.285), `discount_pct` (1.320).
- **Dinheiro (`unit_price`):** mesmos 3 formatos das outras colunas monetárias.
- **Nota conceitual:** `order_id` e `product_id` aqui são chaves estrangeiras — números altos de "duplicados" nessas colunas (11.191 e 25.453) são esperados (muitos itens por pedido, muitos itens do mesmo produto) e não representam sujeira.

### reviews (8.080 linhas)
- **PK duplicada:** 80 `review_id` repetidos (linhas 100% idênticas).
- **FKs íntegras:** nenhum `product_id` ou `customer_id` órfão.
- **Nulos relevantes:** `review_text` (1.630), `rating` (400), `review_date` (391).
- **Datas (`review_date`):** mesma mistura de formatos + timestamp com hora (`08/01/2025 15:21`) + sentinela `0000-00-00`.
- **Outros:** vários `review_text` repetidos (mesmo texto genérico em várias linhas) — não é erro de formato, mas indica que o texto pode ter baixo valor analítico se usado sem cautela.

### Problemas transversais (aparecem em várias tabelas — viram funções reutilizáveis na Silver)
1. **Datas em 5+ formatos** por coluna, todas com o sentinela `0000-00-00` como "nulo disfarçado".
2. **Dinheiro em 3 formatos** (`1234.56` / `"1234,56"` / `"R$ 1.234,56"`) em `price`, `total_amount`, `unit_price`.
3. **Categóricas com variação de caixa** em quase toda coluna categórica (`state`, `status`, `payment_method`, `customer_segment`, `category`).
4. **Mojibake de encoding** isolado em `products.category`.
5. **PKs duplicadas** em 4 das 5 tabelas (só `order_items` está limpa).
6. **Integridade referencial quebrada** em 2 relacionamentos: `orders→customers` (210 órfãos) e `order_items→products` (138 órfãos); `reviews` está 100% íntegra.
7. **Sentinelas de nulo disfarçados de número válido:** `stock_quantity` negativo.

Este mapa é o backlog de decisões da camada **Silver**: cada item acima vira uma regra de limpeza explícita e testável.
